contracts/upgradeable_contracts/BaseMediatorFeeManager.sol

```diff
@@ -1,164 +1,108 @@
-        address[] memory list = new address[](rewardAccountsCount());
-        uint256 counter = 0;
-        address nextRewardAccount = getNextRewardAccount(F_ADDR);
-        require(nextRewardAccount != address(0));
-
-        while (nextRewardAccount != F_ADDR) {
-            list[counter] = nextRewardAccount;
-            nextRewardAccount = getNextRewardAccount(nextRewardAccount);
-            counter++;
-
-            require(nextRewardAccount != address(0));
-        }
-
-        return list;
+    constructor(address _owner, uint256 _fee, address[] _rewardAccountList) public {
+        require(_rewardAccountList.length > 0 && _rewardAccountList.length <= MAX_REWARD_ACCOUNTS);
+        _transferOwnership(_owner);
+        _setFee(_fee);
+        rewardAccounts = _rewardAccountList;
     }
 
     /**
     * @dev Calculates the fee amount to be subtracted from the value.
     * @param _value the base value from which fees are calculated
     */
     function calculateFee(uint256 _value) external view returns (uint256) {
-        uint256 fee = getFee();
         return _value.mul(fee).div(MAX_FEE);
     }
 
     /**
-    * @dev Sets the fee percentage amount for the mediator operations.
+    * @dev Stores the fee percentage amount for the mediator operations.
     * @param _fee the fee percentage
     */
-    function setFee(uint256 _fee) external validFee(_fee) {
-        uintStorage[FEE_STORAGE_KEY] = _fee;
+    function _setFee(uint256 _fee) internal validFee(_fee) {
+        fee = _fee;
         emit FeeUpdated(_fee);
     }
 
     /**
-    * @dev Tells the fee percentage amount for the mediator operations.
-    * @return the fee percentage amount
+    * @dev Sets the fee percentage amount for the mediator operations. Only the owner can call this method.
+    * @param _fee the fee percentage
     */
-    function getFee() public view returns (uint256) {
-        return uintStorage[FEE_STORAGE_KEY];
+    function setFee(uint256 _fee) external onlyOwner {
+        _setFee(_fee);
     }
 
     /**
     * @dev Adds a new account to the list of accounts to receive rewards for the operations.
+    * Only the owner can call this method.
     * @param _account new reward account
     */
-    function addRewardAccount(address _account) external {
-        require(_account != address(0) && _account != F_ADDR);
+    function addRewardAccount(address _account) external onlyOwner {
+        require(_account != address(0));
         require(!isRewardAccount(_account));
-
-        address firstAccount = getNextRewardAccount(F_ADDR);
-        // if list wasn't initialized
-        if (firstAccount == address(0)) {
-            firstAccount = F_ADDR;
-        }
-        setNextRewardAccount(_account, firstAccount);
-        setNextRewardAccount(F_ADDR, _account);
-        setRewardAccountsCount(rewardAccountsCount().add(1));
+        require(rewardAccounts.length.add(1) < MAX_REWARD_ACCOUNTS);
+        rewardAccounts.push(_account);
```

Comment:
Could you provide more info why did you choose to return back to the array instead of the linked list?

Reply:
Since we don't have to use eternal storage in this contract I saw the opportunity to try using an array.
Using an array means only having one variable to store all the data of the list of accounts against using a mapping, a uint variable to store the count and defining a F_ADDR constant. Also, it looks to me that it is less error-prone, it is not needed to build the list manually, it is easier to iterate and overall it results in a little more readable code.

Regarding gas usage, I compared using an array vs a [linkedList](https://gist.github.com/patitonar/fdcff47b13d7936997c95c8a34e30651)

| Method        | Array           | LinkedList |
| ------------- |:-------------:| -----:|
| Deploy (initializes list with 3 accounts)     | 947903 | 1028470 |
| addRewardAccount      | 52482 | 55526 |
| removeRewardAccount | 32351 | 25628 |
| distributeFee | 105385 | 105911 |

The most used method which will be `distributeFee` has a similar gas usage.

What do you think?
I'm OK using a LinkedList which will make sense for consistency since we use it in all of the other contracts.

Reply:
Thanks for the explanation. 

I agree that we can use the array here instead of linked list. I inspected the code of the `BaseMediatorFeeManager` contract and found that random access to any element of the list is not required. The only place where `isRewardAccount` is used can be optimized to reduce number of loop iterations. The operation `removeRewardAccount` is also is not going to be used often so, we can close eyes that its efficiency a bit higher.

I also imagined the bridge where the Fee Manager will be used and found that it does not make sense to spend resources now to carefully introduce and test the linked list:
  * WETC bridge -- we have only one validator. The usage of the linked list here is significant overhead.
  * POA bridge -- it is not clear if the fees will be collected there at all. 
  * STAKE bridge -- the fee will be distributed by using a consensus contract (https://www.staketoken.net/token-model/reward-mechanics), the bridge will have no information about the stakers at all.

That's why we will leave this functionality as is -- with the array of the reward addresses.